src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+            raise TypeError("slot must be a string")
+
+        if "attributes" in task and not isinstance(
+            task["attributes"], collectionsAbc.Mapping
+        ):
+            raise TypeError("count must be a mapping")
+
+        command = task["command"]
+        if not (
+            (  # sequence of strings - N.B. also true for just a plain string
+                isinstance(command, collectionsAbc.Sequence)
+                and all(isinstance(x, six.string_types) for x in command)
+            )
+        ) or isinstance(command, six.string_types):
+            raise TypeError("command must be a list of strings")
+
+    @staticmethod
+    def _validate_attributes(attributes):
+        _validate_keys(["system", "user"], attributes.keys(), keys_optional=True)
+
+    @staticmethod
+    def _create_resource(res_type, count, with_child=None):
+        if with_child is not None and not isinstance(
+            with_child, collectionsAbc.Sequence
+        ):
+            raise TypeError("child resource must None or a sequence")
+        elif with_child is not None and isinstance(with_child, six.string_types):
+            raise TypeError("child resource must not be a string")
+        if not count > 0:
+            raise ValueError("resource count must be > 0")
+
+        res = {"type": res_type, "count": count}
+
+        if with_child:
+            res["with"] = with_child
+        return res
+
+    @classmethod
+    def _create_slot(cls, label, count, with_child):
+        slot = cls._create_resource("slot", count, with_child)
+        slot["label"] = label
+        return slot
+
+    @property
+    def duration(self):
+        try:
+            return self.jobspec["attributes"]["system"]["duration"]
+        except KeyError:
+            return None
+
+    @duration.setter
+    def duration(self, duration):
+        """
+        Assign a time limit to the job.  The duration may be:
+        - a float in seconds
+        - a string in Flux Standard Duration
+        A duration of zero is interpreted as "not set".
+        """
+        if isinstance(duration, six.string_types):
+            time = parse_fsd(duration)
+        elif isinstance(duration, float):
+            time = duration
+        else:
+            raise ValueError("duration must be a float or string")
+        if time < 0:
+            raise ValueError("duration must not be negative")
+        if math.isnan(time) or math.isinf(time):
+            raise ValueError("duration must be a normal, finite value")
+        self.jobspec["attributes"]["system"]["duration"] = time
+
+    @property
+    def cwd(self):
+        """
+        Get working directory of job.
+        """
+        try:
+            self.jobspec["attributes"]["system"]["cwd"]
+        except KeyError:
+            return
+
+    @cwd.setter
+    def cwd(self, cwd):
+        """
+        Set working directory of job.
+        """
+        if not isinstance(cwd, six.string_types):
```

Comment:
This is another one where the python standard type would probably be a good addition, a 3.4+ pathlib Path object specifically.